net/udprelay/server.go

```diff
@@ -337,37 +338,49 @@ func NewServer(logf logger.Logf, port uint16, onlyStaticAddrPorts bool) (s *Serv
 		Logf:   logger.WithPrefix(logf, "netcheck: "),
 		SendPacket: func(b []byte, addrPort netip.AddrPort) (int, error) {
 			if addrPort.Addr().Is4() {
-				return s.uc4.WriteToUDPAddrPort(b, addrPort)
-			} else if s.uc6 != nil {
-				return s.uc6.WriteToUDPAddrPort(b, addrPort)
+				return s.uc4[0].WriteToUDPAddrPort(b, addrPort)
+			} else if len(s.uc6) > 0 {
+				return s.uc6[0].WriteToUDPAddrPort(b, addrPort)
 			} else {
 				return 0, errors.New("IPv6 socket is not bound")
 			}
 		},
 	}
 
-	err = s.listenOn(port)
+	err = s.bindSockets(port)
 	if err != nil {
 		return nil, err
 	}
+	s.startPacketReaders()
 
 	if !s.onlyStaticAddrPorts {
 		s.wg.Add(1)
 		go s.addrDiscoveryLoop()
 	}
 
-	s.wg.Add(1)
-	go s.packetReadLoop(s.uc4, s.uc6, true)
-	if s.uc6 != nil {
-		s.wg.Add(1)
-		go s.packetReadLoop(s.uc6, s.uc4, false)
-	}
 	s.wg.Add(1)
 	go s.endpointGCLoop()
 
 	return s, nil
 }
 
+func (s *Server) startPacketReaders() {
+	for _, readSockets := range []*[]batching.Conn{&s.uc4, &s.uc6} {
```

Comment:
I feel like this might be easier to read if we did two serial loops rather than nested.

```go
for i, uc := range s.uc4 {
   var other batching.Conn
   if len(s.uc6) {
      other = s.ut6[min(len(s.uc6)-1, i)]
   }
   s.wg.Go(func() { s.packetReadLoop(uc, other, true) })
}
for i, uc := range s.uc6 {
   var other batching.Conn
   if len(s.uc4) {
      other = s.ut6[min(len(s.uc4)-1, i)]
   }
   s.wg.Go(func() { s.packetReadLoop(uc, other, false) })
}
```

Acknowledged, it's a little duplicative of structure, but less pointer indirection (which I think detracts from clarity here) and no need for the swap condition. And since it all lives right here, it won't be too likely to diverge.